### ISSUE TITLE
How to handle Ctrl-D?

### ISSUE THREAD

**Issue by majewsky** [vt6/vt6#26]
It's next to impossible to convince the kernel to signal EOF on a readable FD with a not-yet-closed writing counterpart (be it a pipe or a socket or whatever) unless that FD is a tty slave.

**Comment by majewsky** [comment]
This works (run with strace):

```c
#include <string.h>
#include <strings.h>
#include <sys/socket.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main() {
  int fds[2];
  if (socketpair(AF_UNIX, SOCK_SEQPACKET, 0, fds) < 0) {
    perror("socketpair");
    return 1;
  }

  char buf[1024];
  strcpy(buf, "hallo");
  if (send(fds[0], buf, strlen(buf), 0) < 0) {
    perror("write");
    return 1;
  }
  char nix[] = { '\0' };
  if (send(fds[0], nix, 0, MSG_EOR) < 0) {
    perror("write");
    return 1;
  }
  if (send(fds[0], buf, strlen(buf), 0) < 0) {
    perror("write");
    return 1;
  }

  for (int i = 0; i < 5; ++i) {
    bzero(buf, strlen(buf));
    if (read(fds[1], buf, sizeof(buf)) < 0) {
      perror("read");
      return 1;
    }
  }
}
```

**Comment by majewsky** [comment]
After some further poking, it appears that SOCK_SEQPACKET is unsuitable for us. The following amended example...

```c
#include <string.h>
#include <sys/socket.h>
#include <stdio.h>
#include <unistd.h>

int main() {
  int fds[2];
  if (socketpair(AF_UNIX, SOCK_SEQPACKET, 0, fds) < 0) {
    perror("socketpair");
    return 1;
  }

  const char* msgs[] = { "hallo", "", "hallo", NULL };
  for (const char** msg = msgs; *msg; ++msg) {
    if (send(fds[0], *msg, strlen(*msg), 0) < 0) {
      perror("write");
      return 1;
    }
  }

  char buf[4];
  for (;;) {
    if (read(fds[1], buf, sizeof(buf)) < 0) {
      perror("read");
      return 1;
    }
  }
}
```

...behaves as follows: (excerpt from strace)

```strace
socketpair(AF_UNIX, SOCK_SEQPACKET, 0, [3, 4]) = 0
sendto(3, "hallo", 5, 0, NULL, 0)       = 5
sendto(3, "", 0, 0, NULL, 0)            = 0
sendto(3, "hallo", 5, 0, NULL, 0)       = 5
read(4, "hall", 4)                      = 4
read(4, "", 4)                          = 0
read(4, "hall", 4)                      = 4
read(4, ^C
```

The key difference here is that the read buffer is too small to contain the entire messages, so `read()` discards the unread portion of the message. Data loss is of course unacceptable for stdin.

**Comment by majewsky** [comment]
This is resolved by #39 pending the addition of an event, e.g. `term1.eof`.
